### PR TITLE
Trigonometric solvers

### PR DESCRIPTION
**Pull request by ranjithkumar007**
TODO : 
- [x] Fix [failing tests](https://github.com/symengine/symengine/pull/1305/files#diff-b7a4ace1c044cb54270e3f95c63d5291R495).
I will add trig simplification in a different PR.

### CODE REVIEW

**Comment by ranjithkumar007** [comment]
subs for ImageSet and Union are needed for replacing dummy `n` with a symbol.
for other types of sets, we can add `subs` later when required.
Replacing dummy `n` is not needed actually. But without it, I am not able to test the solutions as the dummy created inside `solve` and `tests` are always different(reason for tests failing earlier).
thoughts @srajangarg @Sumith1896 ?

**Comment by ranjithkumar007** [comment]
would you mind reviewing this @srajangarg?

**Comment by ranjithkumar007** [comment]
One test doesn't pass because `expand(I*(x + y))` doesn't simplify. (unrelated to this PR)
I will try to fix it sometime later.

File: symengine/real_imag.cpp [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+            magn = pow(magn, exp_);
+            ang = mul(ang, exp_);
+            *real_ = mul(magn, cos(ang));
+            *imag_ = mul(magn, sin(ang));
+        } else {
+            throw SymEngineException("Not Implemented");
+        }
+    }
+
+    void bvisit(const Complex &x)
```
Comment (isuruf) [inline symengine/real_imag.cpp]: This should be `ComplexBase`

File: symengine/real_imag.cpp [hunk 2]
```diff
@@ -0,0 +1,133 @@
+#include <symengine/visitor.h>
```
Comment (isuruf) [inline symengine/real_imag.cpp]: Rename this file to `as_real_imag.cpp`

**Comment by isuruf** [comment]
Can you send smaller PRs please? This PR has lots of independent stuff.

File: symengine/real_imag.cpp (hunk 1 above)
Comment (isuruf) [inline symengine/real_imag.cpp]: Also need to handle `zoo`

File: symengine/real_imag.cpp [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+        }
+    }
+
+    void bvisit(const Add &x)
+    {
+        RCP<const Basic> re = zero, re_, im = zero, im_;
+
+        for (const auto &arg : x.get_args()) {
+            as_real_imag(arg, outArg(re_), outArg(im_));
+            re = add(re, re_);
```
Comment (isuruf) [inline symengine/real_imag.cpp]: This is inefficient. Use `Add::from_dict` and helpers there.

File: symengine/real_imag.cpp [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+        for (const auto &arg : x.get_args()) {
+            as_real_imag(arg, outArg(re_), outArg(im_));
+            if (eq(*im_, *zero)) {
+                coef = mul(coef, re_);
+            } else if (eq(*re_, *zero)) {
+                is_imag = !is_imag;
+                coef = mul(coef, im_);
+            } else {
+                rest = mul(rest, arg);
```
Comment (isuruf) [inline symengine/real_imag.cpp]: Why?

File: symengine/real_imag.cpp [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+    }
+
+    void bvisit(const Symbol &x)
+    {
+        throw SymEngineException("Not Implemented");
+    }
+
+    void bvisit(const Basic &x)
+    {
+        *real_ = x.rcp_from_this();
```
Comment (isuruf) [inline symengine/real_imag.cpp]: What about trig functions?

**Comment by ranjithkumar007** [comment]
Okay! I will make smaller PRs.

File: symengine/functions.cpp [hunk 6]
```diff
@@ -3571,4 +3620,23 @@ RCP<const Basic> min(const vec_basic &arg)
     }
 }
 
+RCP<const Basic> expand_as_exp(const RCP<const Basic> &self)
+{
+    map_basic_basic mp_args;
+    for (const auto &a : self->get_args()) {
+        mp_args[a] = expand_as_exp(a);
+    }
```
Comment (ranjithkumar007) [inline symengine/functions.cpp]:
```
template <typename Func>
RCP<const Basic> bottom_up(const RCP<const Basic> &barg, const Func &f)
{
    if (is_a<Add>(*barg) or is_a<Mul>(*barg)) {
        vec_basic newargs;
        for (const auto &a : barg->get_args()) {
            newargs.push_back(bottom_up(a, f));
        }
        if (is_a<Add>(*barg))
            return f(add(newargs));
        return f(mul(newargs));
    }
    if (is_a<Pow>(*barg)) {
        auto &dfunc = down_cast<const Pow &>(*barg);
        auto farg1 = dfunc.get_base(), farg2 = dfunc.get_exp();
        auto newarg1 = bottom_up(farg1, f), newarg2 = bottom_up(farg2, f);
        auto nbarg = barg;
        if (farg1 != newarg1 or farg2 != newarg2)
            nbarg = pow(newarg1, newarg2);
        return f(nbarg);
    }
    if (is_a_sub<OneArgFunction>(*barg)) {
        const OneArgFunction &dfunc = down_cast<const OneArgFunction &>(*barg);
        auto farg = dfunc.get_arg();
        auto newarg = bottom_up(farg, f);
        auto nbarg = barg;
        if (not eq(*newarg, *farg))
            nbarg = dfunc.create(newarg);
        return f(nbarg);
    }
    if (is_a_sub<TwoArgFunction>(*barg)) {
        const TwoArgFunction &dfunc = down_cast<const TwoArgFunction &>(*barg);
        auto farg1 = dfunc.get_arg1(), farg2 = dfunc.get_arg2();
        auto newarg1 = bottom_up(farg1, f), newarg2 = bottom_up(farg2, f);
        auto nbarg = barg;
        if (farg1 != newarg1 or farg2 != newarg2)
            nbarg = dfunc.create(newarg1, newarg2);
        return f(nbarg);
    }
    if (is_a_sub<MultiArgFunction>(*barg)) {
        const MultiArgFunction &dfunc
            = down_cast<const MultiArgFunction &>(*barg);
        auto fargs = dfunc.get_args();
        vec_basic newargs;
        for (const auto &a : fargs) {
            newargs.push_back(bottom_up(a, f));
        }
        auto nbarg = dfunc.create(newargs);
        return f(nbarg);
    }
    return barg;
}

RCP<const Basic> expand_as_exp(const RCP<const Basic> &self)
{
    // map_basic_basic mp_args;
    // for (const auto &a : self->get_args()) {
    //     mp_args[a] = expand_as_exp(a);
    // }

    static auto f = [](const RCP<const Basic> &a) -> RCP<const Basic> {
        if (is_a<Sin>(*a) or is_a<Cos>(*a) or is_a<Tan>(*a)
            or is_a<Cot>(*a) or is_a<Sec>(*a) or is_a<Csc>(*a)
            or is_a<Sinh>(*a) or is_a<Cosh>(*a) or is_a<Tanh>(*a)
            or is_a<Coth>(*a) or is_a<Sech>(*a)
            or is_a<Csch>(*a)) // expandables
        {
            const OneArgFunction &func = down_cast<const OneArgFunction &>(*a);
            return func.expand_as_exp();
        }
        return a;
    };
    return bottom_up(self, f);
    // return self->subs(mp_args);
}
```

@isuruf Is this the efficient way?

Comment (isuruf) [inline symengine/functions.cpp]: Change `bottom_up` to a `Visitor` class and extend that for `expand_as_exp` by overriding the `void bvisit(const Sin &)` method and other methods.

Comment (ranjithkumar007) [inline symengine/functions.cpp]: Thanks :)

**Comment by ranjithkumar007** [comment]
@srajangarg, can you please review this?

**Comment by ranjithkumar007** [comment]
Any idea on why test cases are failing on `MSVC15` compiler [here](https://ci.appveyor.com/project/certik/symengine/build/3006)?

**Comment by srajangarg** [comment]
Can you pinpoint which tests fail?

**Comment by ranjithkumar007** [comment]
```
C:\projects\symengine\symengine\tests\basic\test_solve.cpp(388): FAILED:
  REQUIRE( ! is_a_LinearArgTrigEquation(*add(tan(x), x), *x) )
with expansion:
  false
```

**Comment by ranjithkumar007** [comment]
tests passed on `MSVC15`. 
ping @srajangarg

**Comment by ranjithkumar007** [comment]
https://travis-ci.org/symengine/symengine/jobs/259971628#L9153 shows some weird error message.

**Comment by srajangarg** [comment]
@isuruf can you review too?

**Comment by srajangarg** [comment]
`g++-4.8: internal compiler error: Killed (program cc1plus)` @isuruf compiler bug?

**Comment by isuruf** [comment]
Travis-ci might have killed the compiler if it was taking too much memory.

**Comment by ranjithkumar007** [comment]
@isuruf, how can I fix this memory issue?

**Comment by ranjithkumar007** [comment]
restarting [this](https://travis-ci.org/ranjithkumar007/symengine/jobs/259971582) build on my fork ended up with a different [error](https://travis-ci.org/ranjithkumar007/symengine/jobs/259971582#L9135)
`mktemp: failed to create file via template ‘/tmp/codecov.adjustments.XXXXXX’: No space left on device` .

**Comment by isuruf** [comment]
Can you try removing Piranha from the test?

File: symengine/subs.h [hunk 7]
```diff
@@ -99,10 +99,25 @@ class SubsVisitor : public BaseVisitor<SubsVisitor>
     {
         RCP<const Basic> base_new = apply(x.get_base());
         RCP<const Basic> exp_new = apply(x.get_exp());
-        if (base_new == x.get_base() and exp_new == x.get_exp())
+        if (subs_dict_.size() == 1 and is_a<Pow>(*((*subs_dict_.begin()).first))
+            and not is_a<Add>(
+                    *down_cast<const Pow &>(*(*subs_dict_.begin()).first)
+                         .get_exp())) {
```
Comment (isuruf) [inline symengine/subs.h]: At the moment, `subs` in symengine is like `xreplace` in sympy and doesn't do these kind of simplifications. If you want to do this, I would rename the existing method to `xreplace` (`XReplaceVisitor`) and then use `subs` (`SubsVisitor` inheriting from `XReplaceVisitor`) for the new implementation.

Comment (ranjithkumar007) [inline symengine/subs.h]: Never noticed this before.
Thanks.  will update this asap.

Comment (ranjithkumar007) [inline symengine/subs.h]: what about `msubs`? Should that be replaced with `mxreplace`?

Comment (ranjithkumar007) [inline symengine/subs.h]: Also, I think we need to add a class for `XReplace` similar to `Subs`?

Comment (isuruf) [inline symengine/subs.h]: `msubsvisitor` should derive from `xreplacevisitor`.

`Also, I think we need to add a class for XReplace similar to Subs?` No need. `Subs` is enough.

Comment (isuruf) [inline symengine/subs.h]: Also, a separate PR would be great for this.

**Comment by ranjithkumar007** [comment]
@srajangarg, I tried moving piranha to a different test here. This landed in a different issue 
```
CMake Error at symengine/cmake_install.cmake:36 (file):
  file INSTALL cannot copy file
  "/home/travis/build/symengine/symengine/build/symengine/libsymengine.a" to
  "/home/travis/our_usr/lib/libsymengine.a".
Call Stack (most recent call first):
  cmake_install.cmake:63 (include)
```

File: .travis.yml [hunk 8]
```diff
@@ -93,7 +93,17 @@ matrix:
...
+      compiler: gcc
+      os: linux
+      addons:
+        apt:
+          sources:
+          - ubuntu-toolchain-r-test
+          packages:
+          - binutils-dev
+          - g++-4.7
+    - env: BUILD_TYPE="Debug" WITH_BFD="yes" WITH_PIRANHA="yes" WITH_COVERAGE="yes" MAKEFLAGS="-j2"
```
Comment (isuruf) [inline .travis.yml]: Problem was with piranha and coverage. This changes nothing

File: bin/install_travis.sh [hunk 9]
```diff
@@ -90,7 +90,7 @@ if [[ "${WITH_MPC}" == "yes" ]]; then
 fi
 
 if [[ "${WITH_FLINT}" == "yes" ]] && [[ "${WITH_FLINT_DEV}" != "yes" ]]; then
-    conda_pkgs="$conda_pkgs libflint=2.5.2"
+    conda_pkgs="$conda_pkgs libflint=2.5.2 cmake=3.6.2"
```
Comment (isuruf) [inline bin/install_travis.sh]: Why?

File: .travis.yml (hunk 8 above)
Comment (ranjithkumar007) [inline .travis.yml]: Should I disable coverage here?

Comment (isuruf) [inline .travis.yml]: Yes, also don't add a new test. add `WITH_PIRANHA=yes` to an existing test like https://github.com/symengine/symengine/blob/master/.travis.yml#L86

**Comment by ranjithkumar007** [comment]
Finally Travis Passes.
ping @isuruf

**Comment by ranjithkumar007** [comment]
@nishnik can you have a look?

**Comment by ranjithkumar007** [comment]
@isuruf , can you please have a look?

File: symengine/visitor.cpp [hunk 10]
```diff
@@ -175,4 +177,195 @@ void TransformVisitor::bvisit(const MultiArgFunction &x)
     auto nbarg = x.create(newargs);
     result_ = nbarg;
 }
+
+class IsALinearArgTrigVisitor : public BaseVisitor<IsALinearArgTrigVisitor>
```
Comment (isuruf) [inline symengine/visitor.cpp]: Use `StopVisitor` and `preorder_traversal_stop` here.

Comment (ranjithkumar007) [reply symengine/visitor.cpp]: `StopVisitor` is not useful for stopping only locally. What I mean by this, is that If I want to break without seeing arguments of an `expr`, I can do only a global break, which might not be what is desired.
ex. - tan(x) + x
Add(Tan(x), x)
So, After checking tan(x), If I want to skip its arguments and go on to next Add argument I won't be able to do it using these two.

Comment (ranjithkumar007) [reply symengine/visitor.cpp]: Done.

Comment (ranjithkumar007) [reply symengine/visitor.cpp]: @isuruf , I need to skip calling `visit` on arguments of Trig functions like 
`sin(add(x, y))` , Once I see `sin` with arguments linear in `x`, I should skip the preorder_traversal in only that subtree and continue to the next subtree. How can I do this?

Comment (ranjithkumar007) [reply symengine/visitor.cpp]: I introduced a `preorder_traversal_local_stop` which can skip traversing the subtree generated by its arguments. Please let me know if that's Ok.

**Comment by ranjithkumar007** [comment]
ping @certik, @isuruf

File: symengine/visitor.cpp [hunk 11]
```diff
@@ -175,4 +177,185 @@ void TransformVisitor::bvisit(const MultiArgFunction &x)
...
+    if (v.stop_ or v.local_stop_)
+        return;
+    for (const auto &p : b.get_args()) {
+        preorder_traversal_local_stop(*p, v);
+        if (v.stop_)
+            return;
+    }
+}
+
+class IsALinearArgTrigVisitor
```
Comment (isuruf) [inline symengine/visitor.cpp]: Can you move this to `solve.cpp`

Comment (ranjithkumar007) [reply symengine/visitor.cpp]: cool. 
`IsALinearArgTrigVisitor` might not be used in any other situation, Is this the reason for moving?

Comment (isuruf) [reply symengine/visitor.cpp]: Yes, seems like it's needed for `solvers.cpp` only

**Comment by ranjithkumar007** [comment]
updated @isuruf .

File: symengine/solve.h [hunk 12]
```diff
@@ -46,6 +50,13 @@ RCP<const Set> solve_poly_quartic(const vec_basic &coeffs,
                                   const RCP<const Set> &domain
                                   = universalset());
 
+bool is_a_LinearArgTrigEquation(const Basic &b, const Symbol &x);
+
+RCP<const Set> invertComplex(const RCP<const Basic> &fX,
```
Comment (isuruf) [inline symengine/solve.h]: Can you add a comments here about what this method does? It'd be great to have it for all the methods.

**Comment by ranjithkumar007** [comment]
@isuruf ping.

File: symengine/solve.h [hunk 13]
```diff
@@ -16,36 +16,69 @@
...
+ * Helper method to decide if solve_trig can solve a particular equation.
+ * Checks the argument of Trigonometric part, and returns false if it is
+ * non-linear;
+ * true otherwise.
+ */
+bool is_a_LinearArgTrigEquation(const Basic &b, const Symbol &x);
+
+// returns Inverse of a complex equation `fX = gY` wrt symbol `sym`.
+// Dummy `nD` is used as the symbol for `ImageSet` while returning the solution
+// set.
```
Comment (isuruf) [inline symengine/solve.h]: I still don't understand what this does.

**Comment by certik** [comment]
@ranjithkumar007 would you mind please rebasing this on top of the latest master and answering @isuruf's comment?

**Comment by certik** [comment]
@ranjithkumar007 ping.

**Comment by certik** [comment]
That looks good, thanks!

**Pull request merged**